### ISSUE TITLE
Include ONNX Runtime in Unreal Engine 4

### ISSUE THREAD

**Issue by dilne** [microsoft/onnxruntime#10669]
Hi, I'd like to include ORT inside UE4. I'm doing this by following the tutorial at https://unreal.blog/how-to-include-any-third-party-library and using [onnxruntime-win-x64-1.10.0.zip](https://github.com/microsoft/onnxruntime/releases/download/v1.10.0/onnxruntime-win-x64-1.10.0.zip).

I get build errors after following the tutorial. Is implementing ORT even possible? The tutorial says a static library is required. Any help is much appreciated, even if it is just to point me in the right direction. Thank you

**Comment by chilo-ms** [comment]
Leveraging ORT is definitely possible as long as you have ORT header files and libraries (onnxruntime-win-x64-1.10.zip should be working). You probably need to revisit the build instructions and error messages to see what the problem really is.

**Comment by dilne** [comment]
Thanks @chilo-ms. Any suggestions from anyone on what causes this build error are much appreciated.

To quickly detail my process, I add `PrivateIncludePathModuleNames.AddRange(new string[] { "ORT" });` to my project's build.cs file, thus, enabling the plugin. I have added the include headers, .libs, and .dlls from the ORT zip to my plugins folder and added the following to my plugin's build.cs file:

```
PublicIncludePaths.AddRange(
			new string[] {
				"$(PluginDir)\\" + "include"
			}
			);

		PublicAdditionalLibraries.AddRange(
			new string[] {
				"$(PluginDir)\\" + "LibraryBinaries\\" + "Windows\\" + "onnxruntime.lib",
				"$(PluginDir)\\" + "LibraryBinaries\\" + "Windows\\" + "onnxruntime_providers_shared.lib"
			}
			);

		PublicDelayLoadDLLs.AddRange(
			new string[] {
				"onnxruntime.dll",
				"onnxruntime_providers_shared.dll"
			}
			);

		RuntimeDependencies.Add("$(PluginDir)\\LibraryBinaries\\Windows\\onnxruntime.dll");
		RuntimeDependencies.Add("$(PluginDir)\\LibraryBinaries\\Windows\\onnxruntime_providers_shared.dll");
```

It builds correctly, but when I add `#include <onnxruntime_cxx_api.h>` to my main .cpp I get the following build error:


```
Build started...
1>------ Build started: Project: ImplementingONNX, Configuration: Development_Editor x64 ------
1>Building ImplementingONNXEditor...
1>Using Visual Studio 2019 14.29.30140 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133) and Windows 10.0.19041.0 SDK (C:\Program Files (x86)\Windows Kits\10).
1>Building 4 actions with 24 processes...
1>  [1/4] TestingONNX.cpp
1>C:\Users\dmiln\UE4\ImplementingONNX\Plugins\ORT\include\onnxruntime_c_api.h(121): warning C4005: 'ORT_API': macro redefinition
1>  C:\Users\dmiln\UE4\ImplementingONNX\Intermediate\Build\Win64\UE4Editor\Development\ImplementingONNX\Definitions.ImplementingONNX.h(289): note: see previous definition of 'ORT_API'
1>  [2/4] UE4Editor-ImplementingONNX.lib
1>     Creating library C:\Users\dmiln\UE4\ImplementingONNX\Intermediate\Build\Win64\UE4Editor\Development\ImplementingONNX\UE4Editor-ImplementingONNX.lib and object C:\Users\dmiln\UE4\ImplementingONNX\Intermediate\Build\Win64\UE4Editor\Development\ImplementingONNX\UE4Editor-ImplementingONNX.exp
1>  [3/4] UE4Editor-ImplementingONNX.dll
1>     Creating library C:\Users\dmiln\UE4\ImplementingONNX\Intermediate\Build\Win64\UE4Editor\Development\ImplementingONNX\UE4Editor-ImplementingONNX.suppressed.lib and object C:\Users\dmiln\UE4\ImplementingONNX\Intermediate\Build\Win64\UE4Editor\Development\ImplementingONNX\UE4Editor-ImplementingONNX.suppressed.exp
1>TestingONNX.cpp.obj : error LNK2019: unresolved external symbol OrtGetApiBase referenced in function "void __cdecl `dynamic initializer for 'public: static struct OrtApi const * const Ort::Global<void>::api_''(void)" (??__E?api_@?$Global@X@Ort@@2PEBUOrtApi@@EB@@YAXXZ)
1>C:\Users\dmiln\UE4\ImplementingONNX\Binaries\Win64\UE4Editor-ImplementingONNX.dll : fatal error LNK1120: 1 unresolved externals
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets(45,5): error MSB3073: The command ""C:\Program Files\Epic Games\UE_4.27\Engine\Build\BatchFiles\Build.bat" ImplementingONNXEditor Win64 Development -Project="C:\Users\dmiln\UE4\ImplementingONNX\ImplementingONNX.uproject" -WaitMutex -FromMsBuild" exited with code 6.
1>Done building project "ImplementingONNX.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```
The code on line 289 signified in the build error above is: `#define ORT_API DLLIMPORT`
The code at (45,5) signified in the build error above is: `<Exec Command="$(NMakeBuildCommandLine)" Condition="'$(NMakeUseOemCodePage)' == 'true' and '$(NMakeBuildCommandLine)'!=''"/>`

I've tried the suggestions at https://github.com/microsoft/onnxruntime/issues/2081, but these don't help
Any suggestions on how to fix this are greatly appreciated, thanks so much

**Comment by chilo-ms** [comment]
Just saw the important notes from UE4 tutorial: 
The library should be static, not shared or dynamic, and consequently, should not have one of these extensions: "*.dll, *.so, *.dylib".

The onnxruntime library in onnxruntime-win-x64-1.10.zip is shared library. You may need to build from source (don't specify "--build_shared_lib") to get static library.

**Comment by chilo-ms** [comment]
Just saw the important notes from UE4 tutorial:
The library should be static, not shared or dynamic, and consequently, should not have one of these extensions: "*.dll, *.so, *.dylib".

If that's the case, you may need to build onnxruntime from source (don't specify "--build_shared_lib") to get static library.
The onnxruntime library in onnxruntime-win-x64-1.10.zip is shared library.

**Comment by dilne** [comment]
I tried making a static library, but do you know where the .lib files and headers go once the build is complete? I made a new issue asking this in more detail here: https://github.com/microsoft/onnxruntime/issues/10751

**Comment by RyanUnderhill** [comment]
Looks like this was handled in #10751
